From dd60ea7312c7316d39cdf4f29ffb7dd38b455af7 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Sat, 24 Jan 2026 00:15:44 +0100
Subject: [PATCH 2/3] HACK: kernel32: Lock GetProcAddress hack to when needed.

---
 dlls/kernel32/module.c | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/dlls/kernel32/module.c b/dlls/kernel32/module.c
index 54cf01e04ff..d5e81c49adc 100644
--- a/dlls/kernel32/module.c
+++ b/dlls/kernel32/module.c
@@ -263,7 +263,24 @@ BOOL WINAPI GetBinaryTypeA( LPCSTR lpApplicationName, LPDWORD lpBinaryType )
 }
 
 #ifdef __x86_64__
-/* workaround for tpshell */
+static BOOL needs_int3_hack(void)
+{
+    static volatile char cache = -1;
+    BOOL ret = cache;
+    if (ret == -1)
+    {
+        const WCHAR *p, *name = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
+        if ((p = wcsrchr(name, '/')))
+            name = p + 1;
+        if ((p = wcsrchr(name, '\\')))
+            name = p + 1;
+        
+        ret = (!wcsicmp(name, L"Endfield.exe"));
+        cache = ret;
+    }
+    return ret;
+}
+
 static void __attribute__((naked)) int3_stub( void )
 {
     asm("int3\t\n"
@@ -297,7 +314,7 @@ FARPROC get_proc_address( HMODULE hModule, LPCSTR function )
         ANSI_STRING     str;
 
 #ifdef __x86_64__
-        if (strcmp( function, "KiUserApcDispatcher" ) == 0 || strcmp( function, "KiUserCallbackDispatcher" ) == 0)
+        if (needs_int3_hack() && (strcmp( function, "KiUserApcDispatcher" ) == 0 || strcmp( function, "KiUserCallbackDispatcher" ) == 0))
         {
             FIXME( "HACK: returning int3 stub instead of %s\n", function );
             return (FARPROC)&int3_stub;
-- 
2.52.0

